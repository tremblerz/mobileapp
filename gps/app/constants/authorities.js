export const PUBLIC_DATA_URL =
  'https://raw.githubusercontent.com/beoutbreakprepared/nCoV2019/master/latest_data/latestdata.csv';

export const AUTHORITIES_LIST_URL =
  'https://raw.githubusercontent.com/Path-Check/safeplaces-frontend/develop/healthcare-authorities.yaml';
