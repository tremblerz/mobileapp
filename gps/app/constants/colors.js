const colors = {
  BLACK: '#000',
  WHITE: '#FFF',
  DODGER_BLUE: '#428AF8',
  SILVER: '#BEBEBE',
  TORCH_RED: '#F8262F',
  MISCHKA: '#E5E4E6',
  PRIMARY_TEXT: '#000',
  SUCCESS: '#41dca4', // Green
  WARNING: '#ffc000', // Orange
  VIOLET_ALPHA_06: '#f7f8ff', // cannot use transparency
  VIOLET: '#4051DB',
  TRANSPARENT: 'transparent',

  // For notifications
  HIGHEST_RISK: '#e74c3c',
  MIDDLE_RISK: '#e74c3c',
  LOWER_RISK: '#e74c3c',
  LOWEST_RISK: '#e74c3c',

  REG_BUTTON: '#428AF8',
  POS_BUTTON: '#32A852',
  NEG_BUTTON: '#F8262F',
  SENS_BUTTON: '#6C3794',
  VIOLET_BUTTON: '#6979F8',
  VIOLET_BUTTON_DARK: '#3A4CD7',

  VIOLET_TEXT: '#3C4ED8',

  GRAY_BACKGROUND: '#DADADA',

  INTRO_WHITE_BG: '#F7F8FF',
  PULSE_WHITE: '#FFFFFF40',

  DIVIDER: '#A5AFFB',

  BLUE_BUTTON: '#05369B',
  BLUE_TO_BUTTON: '#185BD3',
  RED_BUTTON: '#E3554E',
  RED_TO_BUTTON: '#E97C79',
  GRAY_BUTTON: '#868686',
  GRAY_TO_BUTTON: '#868686',

  MAP_LINE_STROKE: '#4052DB',

  BOTTOM_SHEET_HEADER_INDICATOR: '#E1E4FF',

  BLUE_LINK: '#007AFF',

  RED_TEXT: '#FF5656',

  NAV_BAR_VIOLET: '#5061E6',
  NAV_BAR_BORDER: 'rgba(189, 195, 199,0.6)',

  DARK_GRAY: '#3C475B',
  LIGHT_GRAY: '#EEEEEE',

  MONO_DARK: '#202020',
  MONO_SECONDARY: '#757677',
};

export default colors;
