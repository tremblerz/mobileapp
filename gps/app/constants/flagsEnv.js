import env from 'react-native-config';

/**
 * Normalizes flags:
 *
 * Example:
 *
 * `{ flag_a: 'true' }` becomes `{ a: true }`
 *
 * @param {{[key: string]: string}} envConfig
 */
export function parseFlags(envConfig) {
  return Object.entries(envConfig)
    .filter(([key]) => key.toLowerCase().startsWith('flag'))
    .reduce((flags, [key, value]) => {
      const flag = key.replace(/^flag_/i, '');

      flags[flag] = value === 'true' || value === '1';
      return flags;
    }, {});
}

export const getBuildtimeFlags = () => parseFlags(env);
