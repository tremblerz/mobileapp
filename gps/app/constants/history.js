/**
 * Max esposure reporting window in days
 */
export const MAX_EXPOSURE_WINDOW_DAYS = 14;

/**
 * The value in minutes of each "bin" in the crossed path data.
 */
export const DEFAULT_EXPOSURE_PERIOD_MINUTES = 5;

/**
 * The value in minutes of how long an exposure at a location is
 *    considered concerning.
 */
export const CONCERN_TIME_WINDOW_MINUTES = 4 * 60; // 4 hours, in minutes

/**
 * The value in minutes of how frequently we should check intersection data if
 *    there has been no change to the authorities
 */
export const MIN_CHECK_INTERSECT_INTERVAL = 6 * 60;

/**
 * The value in minutes for the background task service to register for firing
 */
export const INTERSECT_INTERVAL = 60 * 12; // 12 Hours, the value is received in minutes

/**
 * Format of a single history item
 *
 * @typedef {{
 *   date: import("dayjs").Dayjs,
 *   exposureMinutes: number,
 * }} HistoryDay
 */

/**
 * Exposure history
 *
 * @typedef {!HistoryDay[]} History
 */
