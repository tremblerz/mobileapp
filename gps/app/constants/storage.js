export const LOCATION_DATA = 'LOCATION_DATA';
export const CONTACT_DATA = 'CONTACT_DATA';
export const PARTICIPATE = 'PARTICIPATE';
export const MY_UUIDs = 'MY_UUIDs';
export const CROSSED_PATHS = 'CROSSED_PATHS';
export const LANG_OVERRIDE = 'LANG_OVERRIDE';
export const DEBUG_MODE = 'DEBUG_MODE';
export const AUTHORITY_NEWS = 'AUTHORITY_NEWS';
export const LAST_CHECKED = 'LAST_CHECKED';
export const AUTHORITY_SOURCE_SETTINGS = 'AUTHORITY_SOURCE_SETTINGS';
export const ENABLE_HCA_AUTO_SUBSCRIPTION = 'ENABLE_HCA_AUTO_SUBSCRIPTION';
export const FEATURE_FLAG_VALS = 'FEATURE_FLAG_VALS';
