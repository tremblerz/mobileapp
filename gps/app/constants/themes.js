import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';

import Color from './colors';

/** Violet on pale violet bg. e.g. Settings */
export const defaultTheme = {
  background: Color.VIOLET_ALPHA_06,
  textPrimaryOnBackground: Color.VIOLET,
  textSecondaryOnBackground: 'rgba(64, 81, 219, 0.6)',

  navBar: Color.VIOLET,
  onNavBar: Color.WHITE,

  /** E.g. button bg */
  primary: Color.VIOLET_BUTTON,
  /** E.g. button text color */
  onPrimary: Color.WHITE,

  success: Color.SUCCESS,
  warning: Color.WARNING,
  disabled: 'rgba(64, 81, 219, 0.6)',
  border: Color.DIVIDER,
};

/** White on violet bg. E.g. Main screen */
export const violet = {
  ...defaultTheme,
  background: `linear-gradient(245.21deg, ${Color.VIOLET_BUTTON} 0%, ${Color.VIOLET_BUTTON_DARK} 100%);`,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: 'rgba(255, 255, 255, 0.6)',

  navBarBorder: Color.VIOLET, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.VIOLET,
  disabled: '#ececec',
};

/** White on gray bg. E.g. Possible exposure mode on default screen */
export const charcoal = {
  ...defaultTheme,
  background: Color.DARK_GRAY,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: Color.LIGHT_GRAY,

  navBar: Color.DARK_GRAY,
  navBarBorder: Color.DARK_GRAY, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.DARK_GRAY,

  disabled: '#A9AFBA',
};

const THEME_MAP = {
  default: defaultTheme,
  violet,
  charcoal,
};

/** A view which uses the theme's background color */
export const ThemedBackground = styled.View`
  background-color: ${({ theme }) => theme.background};
`;

/**
 * Return theme config for a known theme name
 *
 * @param {'default' | 'violet' | 'charcoal'} themeName
 */
export const getTheme = themeName => {
  const themeColors = THEME_MAP[themeName];
  if (!themeColors) {
    throw new Error(`Unknown theme ${themeName}`);
  }
  return themeColors;
};

/**
 * Configures themeable sub-components with background, text (primary,
 * secondary), button and other various theme colors.
 *
 * Usage:
 *
 * This will render a charcoal view, with a white heading, and a grey paragraph.
 *
 * ```jsx
 * <Theme use="charcoal" setBackground>
 *   <Typography use="headline2">Test</Typogrphy>
 *   <Typography secondary>Paragraph here</Typogrphy>
 * </Theme>
 * ```
 *
 * It makes available the following `props.theme.x` properties in styled components:
 *
 * - background - the standard bg color
 * - primary - the primary theme/button color on top of bakground (e.g. violet)
 * - onPrimary - color to show on top of primary (usually white)
 * - onPrimary - color to show on top of primary (usually white)
 * - textPrimaryOnBackground - text color to show on background
 * - textSecondaryOnBackground - secondary text color to show on background
 * - warning - warning color (always orange)
 * - success - success color (always green)
 * - border - border/divider color (usually gray)
 *
 * @param {{
 *   use?: string,
 *   setBackground?: boolean,
 * }} param0
 */
export const Theme = ({ use = 'default', setBackground = false, children }) => {
  return (
    <ThemeProvider theme={getTheme(use)}>
      {setBackground ? (
        <ThemedBackground>{children}</ThemedBackground>
      ) : (
        children
      )}
    </ThemeProvider>
  );
};

/**
 * @param {{theme: defaultTheme, secondary?: boolean}} param0
 */
export const themeTextColor = ({ theme, secondary }) =>
  secondary ? theme.textSecondaryOnBackground : theme.textPrimaryOnBackground;
