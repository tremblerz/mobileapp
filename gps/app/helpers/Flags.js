import { assign } from 'lodash';
import React, { useContext, useEffect, useState } from 'react';

import { getBuildtimeFlags } from '../constants/flagsEnv';
import { FEATURE_FLAG_VALS } from '../constants/storage';
import { GetStoreData, SetStoreData } from './General';

/**
 * Converts the flag name from env into a more human readable form.
 *
 * Example:
 *
 * `getCleanedFlagName('hello_world')` becomes `Hello world`
 *
 */
export const getCleanedFlagName = name => {
  const withSpaces = name.replace('_', ' ');
  return (
    withSpaces.charAt(0).toUpperCase() + withSpaces.slice(1, withSpaces.length)
  );
};

export const FlagsContext = React.createContext([
  getBuildtimeFlags(),
  () => {},
]);

/**
 * Overwrites properties of `oldFlags` with properties from `newFlags`,
 * omitting any properties from `newFlags` that don't exist on `oldFlags`
 *
 * Examples:
 *
 * `mergeFlags({feature1: false}, {feature1: true})` becomes `{feature1: true}`
 *
 * `mergeFlags({feature1: false}, {feature1: true, feature2: true})` becomes `{feature1: true}`
 */
export const mergeFlags = (oldFlags, newFlags) => {
  return assign(oldFlags, newFlags, (objectValue, sourceValue) => {
    if (objectValue) {
      return sourceValue;
    }
  });
};

export const FlagsProvider = ({ children }) => {
  const buildTimeFlags = getBuildtimeFlags();
  const [flags, setFlags] = useState(buildTimeFlags);

  useEffect(() => {
    const getInitalFlagVals = async () => {
      const storedFlags = await GetStoreData(FEATURE_FLAG_VALS, false);

      if (storedFlags) {
        // Overwrite existing properties of `getBuildtimeFlags` with stored values from async storage,
        // omitting any stored value that is not present on `getBuildtimeFlags`.
        const initialFlags = mergeFlags(buildTimeFlags, storedFlags);
        setFlags(initialFlags);
        await SetStoreData(FEATURE_FLAG_VALS, initialFlags);
      }
    };

    getInitalFlagVals();
  }, [buildTimeFlags]);

  return (
    <FlagsContext.Provider value={[flags, setFlags]}>
      {children}
    </FlagsContext.Provider>
  );
};

/**
 * Custom hook that can be used to fetch the list of all flags,
 * and to set the value for a particular flag.
 */
export function useFlags() {
  const [flags, setFlags] = useContext(FlagsContext);

  const setFlag = async (flagName, flagVal) => {
    const newFlags = { ...flags, [flagName]: flagVal };
    setFlags(newFlags);
    await SetStoreData(FEATURE_FLAG_VALS, newFlags);
  };

  return [flags, setFlag];
}
