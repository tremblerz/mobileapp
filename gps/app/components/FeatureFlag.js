import { useFlags } from '../helpers/Flags';

/**
 *
 * Usage:
 *
 * ```
 * <FeatureFlag
 *  name="google_import"
 *  fallback={<Text>Old</Text>}
 * >
 *   <FeatureUi />
 * </FeatureFlag>
 * ```
 *
 * @param {{
 *   name: string;
 *   fallback?: import('react').ReactNode;
 *   children: import('react').ReactNode;
 * }} param0
 */
export const FeatureFlag = ({ name, children, fallback }) => {
  const [flags] = useFlags();
  const fallbackRender = fallback || null;

  return flags[name] ? children : fallbackRender;
};
