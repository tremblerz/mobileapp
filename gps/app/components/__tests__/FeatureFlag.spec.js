import { render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';

import { getBuildtimeFlags } from '../../constants/flagsEnv';
import { FlagsProvider } from '../../helpers/Flags';
import { FeatureFlag } from '../FeatureFlag';

jest.mock('../../constants/flagsEnv', () => {
  return {
    getBuildtimeFlags: jest.fn(),
  };
});

it('renders feature if the flag is enabled', () => {
  getBuildtimeFlags.mockReturnValue({ feature1: true });

  const { asJSON } = render(
    <FlagsProvider>
      <FeatureFlag name='feature1'>
        <Text>feature1</Text>
      </FeatureFlag>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('does not render if the flag is disabled', () => {
  getBuildtimeFlags.mockReturnValue({ feature1: false });

  const { asJSON } = render(
    <FlagsProvider>
      <FeatureFlag name='feature2'>
        <Text>feature2</Text>
      </FeatureFlag>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('omits feature if the flag is missing', () => {
  getBuildtimeFlags.mockReturnValue({});
  const { asJSON } = render(
    <FlagsProvider>
      <FeatureFlag name='feature3'>
        <Text>feature3</Text>
      </FeatureFlag>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('renders the fallback instead, if the flag is disabled/omitted', () => {
  getBuildtimeFlags.mockReturnValue({ feature1: false });
  const { asJSON } = render(
    <FlagsProvider>
      <FeatureFlag name='feature2' fallback={<Text>Old UI</Text>}>
        <Text>feature2</Text>
      </FeatureFlag>
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});
