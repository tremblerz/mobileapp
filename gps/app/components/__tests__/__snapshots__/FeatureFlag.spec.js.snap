// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render if the flag is disabled 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`omits feature if the flag is missing 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`renders feature if the flag is enabled 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    feature1
  </Text>
</View>
`;

exports[`renders the fallback instead, if the flag is disabled/omitted 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    Old UI
  </Text>
</View>
`;
